### PR TITLE
ensure moderated file transfers only perform allowed operations

### PR DESCRIPTION
**Pull request by capnspacehook**
Fixes https://github.com/gravitational/teleport-private/issues/1055.
Closes https://github.com/gravitational/teleport-private/issues/1391.

To reduce complexity and the risk of file transfer requests getting reused again (see the above issue 1055) pending file transfer requests are now limited to one per session. Creating multiple in-flight file transfer requests for moderated sessions didn't properly work before anyway, so I don't think we're losing anything by doing this. The SFTP server now also only allows necessary operations for the session's approved file transfer request if there is one present.

This change should be backwards compatible as the API for managing file transfer requests has not changed, only previously non-functional behavior has been removed. Since Nodes re-exec themselves to start an SFTP server, Node processes now sending file transfer request info to SFTP server processes shouldn't pose a compatibility issue.

Reviewed already in https://github.com/gravitational/teleport-private/pull/1399, just need approvals to get this merged.

changelog: only allow necessary operations during moderated file transfers and limit in-flight file transfer requests to one per session

### CODE REVIEW

**Comment by zmb3** [comment]
_improve security_ is a bit vague - can we come up with a better changelog entry?

File: integration/integration_test.go
Comment (zmb3) [inline integration/integration_test.go:7908]: nit: call this `auth` or `authServer`?

File: integration/integration_test.go
Comment (zmb3) [inline integration/integration_test.go:7979]: What does TC mean?

File: integration/integration_test.go
Comment (zmb3) [inline integration/integration_test.go:7987]:
```suggestion
	peerClusterClient, err := peerTC.ConnectToCluster(ctx)
```

File: integration/integration_test.go
Comment (zmb3) [inline integration/integration_test.go:8006]: Why do you ignore errors when closing `peerClusterCli` but check them when closing `peerNodeCli`?

File: integration/integration_test.go
Comment (zmb3) [inline integration/integration_test.go:7998]:
```suggestion
	peerNodeClient, err := peerTC.ConnectToNode(
```

File: integration/integration_test.go
Comment (zmb3) [inline integration/integration_test.go:8010]: All these abbreviations are only saving you 2-3 keystrokes and making things harder to understand IMO.
```suggestion
	peerSession, err := peerSSH.NewSession(ctx)
```

File: integration/integration_test.go
Comment (zmb3) [inline integration/integration_test.go:8026]: This will fail the entire test even if the check fails on the first iteration, when the behavior you want is that it can fail for up to 5 seconds before the test should fail.

Recommend using `require.EventuallyWithT` here.

File: integration/integration_test.go
Comment (zmb3) [inline integration/integration_test.go:8095]: Should we at least capture the first value written to the channel and assert that it is in fact a session join event rather than assuming that's the case?

File: integration/integration_test.go
Comment (zmb3) [inline integration/integration_test.go:8119]:
```suggestion
	err = transferSess.RequestSubsystem(ctx, teleport.SFTPSubsystem)
```

File: integration/integration_test.go
Comment (zmb3) [inline integration/integration_test.go:8171]: I don't see any code here that ignores EOF errors. I see code that ignores _all_ errors, but that's not the behavior that this comment describes.

File: integration/integration_test.go
Comment (zmb3) [inline integration/integration_test.go:8184]: Same as above, use the constant for the SFTP subsystem.

File: lib/srv/regular/sftp.go
Comment (zmb3) [inline lib/srv/regular/sftp.go:136]: Can you explain why we initialize this to a 1-byte array?

File: lib/srv/sess.go
Comment (zmb3) [inline lib/srv/sess.go:444]:
```suggestion
	// If the TELEPORT_MODERATED_SESSION_ID environment variable was not set, return not
```

perhaps?

File: lib/srv/sess.go
Comment (zmb3) [inline lib/srv/sess.go:462]: Why take out a separate lock here? Wouldn't this be simpler if you just deferred the unlock on line 456?

File: lib/srv/sess.go
Comment (zmb3) [inline lib/srv/sess.go:1714]: Should we use `path.Dir` here instead? The `path` package is more specialized for working with paths than the `strings` package.

File: lib/srv/sess.go
Comment (zmb3) [inline lib/srv/sess.go:1735]: If these are file paths the `filepath` package is probably more appropriate.

File: tool/teleport/common/sftp.go
Comment (zmb3) [inline tool/teleport/common/sftp.go:100]: This looks very dangerous to me. If you are reading from a source that has a value of `[]byte{1, 2, 3, 4, 5}`, and the first read returns `[]byte{1, 2}`, you'll silently drop the `2`.

File: tool/teleport/common/sftp.go
Comment (zmb3) [inline tool/teleport/common/sftp.go:111]: Do we need a pointer here? `allowedOps` looks pretty small.

File: tool/teleport/common/sftp.go
Comment (zmb3) [inline tool/teleport/common/sftp.go:121]: When do we plan to address this TODO?

File: tool/teleport/common/sftp.go
Comment (zmb3) [inline tool/teleport/common/sftp.go:166]: These 3 cases look identical? Can we combine them?

```
case methodGet, methodPut, methodSetStat:
```

File: tool/teleport/common/sftp.go
Comment (zmb3) [inline tool/teleport/common/sftp.go:265]: Should we be `trace.Wrap`ping these errors?

File: integration/integration_test.go
Comment (capnspacehook) [reply integration/integration_test.go:7979]: Teleport Client, this is named `tc` in most other tests but I had to differentiate between different clients for different users in this test

File: lib/srv/sess.go
Comment (capnspacehook) [reply lib/srv/sess.go:462]: That's what I originally did, but `s.sessionsMux` can't protect from methods on `session` objects from writing to `sess.fileTransferReq` while we're reading from it here, so I acquire the session mutex to be sure.

File: lib/srv/sess.go
Comment (capnspacehook) [reply lib/srv/sess.go:1735]: SFTP requires paths using back slashes, so `path` makes more sense here.

File: tool/teleport/common/sftp.go
Comment (capnspacehook) [reply tool/teleport/common/sftp.go:82]: I don't see how I could use that here, `bytes.Buffer` can only read from `[]byte` where I need to read from an `io.Reader`

File: tool/teleport/common/sftp.go
Comment (capnspacehook) [reply tool/teleport/common/sftp.go:100]: Unless I'm missing something, `byteReader.buf` always has a size of 1 so a read should never return more than 1 byte. I'll add a check just in case though.

File: tool/teleport/common/sftp.go
Comment (capnspacehook) [reply tool/teleport/common/sftp.go:111]: We don't *need* it, the main reason was to make it easier to check for the zero value or not. I don't think it'd have a major perf or memory impact either way though.

File: tool/teleport/common/sftp.go
Comment (capnspacehook) [reply tool/teleport/common/sftp.go:121]: Soon, it's tracked here: https://github.com/gravitational/teleport-private/issues/1401

File: tool/teleport/common/sftp.go
Comment (capnspacehook) [reply tool/teleport/common/sftp.go:166]: Yes good catch, they were different before but should be combined now.

File: tool/teleport/common/sftp.go
Comment (capnspacehook) [reply tool/teleport/common/sftp.go:265]: These get returned to `github.com/pkg/sftp` code so the trace wouldn't be very useful, I can add them if you think it's still worth doing though.

File: integration/integration_test.go
Comment (capnspacehook) [reply integration/integration_test.go:7908]: Done

File: integration/integration_test.go
Comment (capnspacehook) [reply integration/integration_test.go:7987]: Done

File: integration/integration_test.go
Comment (capnspacehook) [reply integration/integration_test.go:8006]: Ah that was a mistake, corrected

File: integration/integration_test.go
Comment (capnspacehook) [reply integration/integration_test.go:7998]: Done

File: integration/integration_test.go
Comment (capnspacehook) [reply integration/integration_test.go:8026]: Done

File: integration/integration_test.go
Comment (capnspacehook) [reply integration/integration_test.go:8095]: That's probably better, done

File: integration/integration_test.go
Comment (capnspacehook) [reply integration/integration_test.go:8171]: Fair point, updated to actually require nil or EOF errors

File: integration/integration_test.go
Comment (capnspacehook) [reply integration/integration_test.go:8184]: Done

File: lib/srv/sess.go
Comment (capnspacehook) [reply lib/srv/sess.go:444]: Thanks, done

File: lib/srv/sess.go
Comment (capnspacehook) [reply lib/srv/sess.go:1714]: I don't think it matters too much in this case, but I changed these to use `path.Dir` instead.

File: integration/integration_test.go
Comment (capnspacehook) [reply integration/integration_test.go]: Fair enough, `cmdSess` is named such because there's 2 sessions started by the peer user, this one is the session that executed a command. But I can see how that could be a little confusing.

File: lib/srv/regular/sftp.go
Comment (capnspacehook) [reply lib/srv/regular/sftp.go:139]: Added comment that explains this. The file transfer request ends in a null byte so if we don't have any request to send just sending a null byte lets the SFTP process no that there's no file transfer request.

File: integration/integration_test.go
Comment (zmb3) [inline integration/integration_test.go:7987]:
```suggestion
	peerClient, err := instance.NewClient(helpers.ClientConfig{
```

File: integration/integration_test.go
Comment (zmb3) [inline integration/integration_test.go:7977]:
```suggestion
	moderator, err := types.NewUser(modUsername)
```

File: integration/integration_test.go
Comment (zmb3) [inline integration/integration_test.go:8034]:
```suggestion
		assert.NoError(t, err)
```
I think you still need to use `assert` instead of `require` to prevent the whole test from failing (and then return early) but you'll want to double check me on that.

File: lib/srv/sess.go
Comment (zmb3) [inline lib/srv/sess.go:461]:
```suggestion
	// acquire the session mutex lock so sess.fileTransferReq doesn't get
```

File: lib/srv/sess.go
Comment (zmb3) [reply lib/srv/sess.go:463]: Ah, good point. I didn't spot that they were different mutexes.

File: tool/teleport/common/sftp.go
Comment (zmb3) [reply tool/teleport/common/sftp.go:83]: Can we use a `bufio.Reader` in that case? Why do we need to do this without buffering?

Won't reading a JSON-encoded request from a SSH channel one byte at a time be super inefficient?

File: tool/teleport/common/sftp.go
Comment (zmb3) [reply tool/teleport/common/sftp.go:264]: Got it. Sounds fine.

File: lib/srv/sess.go
Comment (capnspacehook) [reply lib/srv/sess.go]: Done

File: integration/integration_test.go
Comment (capnspacehook) [reply integration/integration_test.go]: You're right, I copied a line from above and forgot to change `require` to `assert`, done

File: integration/integration_test.go
Comment (capnspacehook) [reply integration/integration_test.go]: Done

File: integration/integration_test.go
Comment (capnspacehook) [reply integration/integration_test.go]: Done

File: tool/teleport/common/sftp.go
Comment (capnspacehook) [reply tool/teleport/common/sftp.go]: I tried using a `bufio.Reader` before but I didn't continue to read from it when handling the SFTP connection meaning the buffered data was lost. I'm not sure why I didn't think to just use the `bufio.Reader` to read from the connection after the file transfer request was read, but that works.

**Comment by capnspacehook** [comment]
@zmb3 how's the updated changelog entry?

**Review by Tener — APPROVED** [review]

**Review by avatus — APPROVED** [review]

**Comment by capnspacehook** [comment]
@zmb3 can I get another review please?

File: tool/teleport/common/sftp.go
Comment (zmb3) [inline tool/teleport/common/sftp.go:606]: Why do we need `newBufferedReaderCloser()` at all? Does this work?

```suggestion
	bufferedReader := bufio.NewReader(chr)
```

The only thing your custom type does that buffered reader doesn't do is close, but we've already deferred a `chr.Close` above.

**Review by zmb3 — COMMENTED** [review]
We're almost there.

I think we can remove the custom buffered reader type, and I think the changelog entry should mention that only 1 in-flight request is allowed now.

Otherwise looks good!

File: tool/teleport/common/sftp.go
Comment (capnspacehook) [reply tool/teleport/common/sftp.go:606]: `sftp.NewRequestServer` requires an `io.ReadWriteCloser` as it's first parameter, so passing a `bufio.NewReader` won't work as it doesn't implement `io.Closer` as you already said: https://pkg.go.dev/github.com/pkg/sftp@v1.13.6#NewRequestServer

**Comment by capnspacehook** [comment]
@zmb3 the changelog entry look like what you had in mind?

File: tool/teleport/common/sftp.go
Comment (zmb3) [reply tool/teleport/common/sftp.go:606]: Isn't that what io.NopCloser is for?

Comment (capnspacehook) [reply tool/teleport/common/sftp.go:606]: Will do, missed your last comment before I merged

**Pull request merged**
